File: Kernel Simulator/Misc/Writers/FancyWriters/FigletWhereColor.vb
﻿
'    Kernel Simulator  Copyright (C) 2018-2022  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports Figgle

Namespace Misc.Writers.FancyWriters
    Public Module FigletWhereColor

        ''' <summary>
        ''' Writes the figlet text with position support
        ''' </summary>
        ''' <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        ''' <param name="Left">Column number in console</param>
        ''' <param name="Top">Row number in console</param>
        ''' <param name="Return">Whether or not to return to old position</param>
        ''' <param name="FigletFont">Figlet font to use in the text.</param>
        ''' <param name="Vars">Variables to format the message before it's written.</param>
        Public Sub WriteFigletWherePlain(Text As String, Left As Integer, Top As Integer, [Return] As Boolean, FigletFont As FiggleFont, ParamArray Vars() As Object)
            Try
                'Format string as needed
                If Not Vars.Length = 0 Then Text = String.Format(Text, Vars)

                'Write the font
                Text = FigletFont.Render(Text)
                WriteWherePlain(Text, Left, Top, [Return], Vars)
            Catch ex As Exception When Not ex.GetType.Name = "ThreadAbortException"
                WStkTrc(ex)
                KernelError(KernelErrorLevel.C, False, 0, DoTranslation("There is a serious error when printing text."), ex)
            End Try
        End Sub

        ''' <summary>
        ''' Writes the figlet text with position support
        ''' </summary>
        ''' <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        ''' <param name="Left">Column number in console</param>
        ''' <param name="Top">Row number in console</param>
        ''' <param name="Return">Whether or not to return to old position</param>
        ''' <param name="FigletFont">Figlet font to use in the text.</param>
        ''' <param name="ColTypes">A type of colors that will be changed.</param>
        ''' <param name="Vars">Variables to format the message before it's written.</param>
        Public Sub WriteFigletWhere(Text As String, Left As Integer, Top As Integer, [Return] As Boolean, FigletFont As FiggleFont, ColTypes As ColTypes, ParamArray Vars() As Object)
            Try
                'Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                SetConsoleColor(ColTypes)

                'Actually write
                WriteFigletWherePlain(Text, Left, Top, [Return], FigletFont, Vars)
            Catch ex As Exception When Not ex.GetType.Name = "ThreadAbortException"
                WStkTrc(ex)
                KernelError(KernelErrorLevel.C, False, 0, DoTranslation("There is a serious error when printing text."), ex)
            End Try
        End Sub

        ''' <summary>
        ''' Writes the figlet text with position support
        ''' </summary>
        ''' <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        ''' <param name="Left">Column number in console</param>
        ''' <param name="Top">Row number in console</param>
        ''' <param name="Return">Whether or not to return to old position</param>
        ''' <param name="FigletFont">Figlet font to use in the text.</param>
        ''' <param name="colorTypeForeground">A type of colors that will be changed for the foreground color.</param>
        ''' <param name="colorTypeBackground">A type of colors that will be changed for the background color.</param>
        ''' <param name="Vars">Variables to format the message before it's written.</param>
        Public Sub WriteFigletWhere(Text As String, Left As Integer, Top As Integer, [Return] As Boolean, FigletFont As FiggleFont, colorTypeForeground As ColTypes, colorTypeBackground As ColTypes, ParamArray Vars() As Object)
            Try
                'Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                SetConsoleColor(colorTypeForeground)
                SetConsoleColor(colorTypeBackground, True)

                'Actually write
                WriteFigletWherePlain(Text, Left, Top, [Return], FigletFont, Vars)
            Catch ex As Exception When Not ex.GetType.Name = "ThreadAbortException"
                WStkTrc(ex)
                KernelError(KernelErrorLevel.C, False, 0, DoTranslation("There is a serious error when printing text."), ex)
            End Try
        End Sub

        ''' <summary>
        ''' Writes the figlet text with position support
        ''' </summary>
        ''' <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        ''' <param name="Left">Column number in console</param>
        ''' <param name="Top">Row number in console</param>
        ''' <param name="Return">Whether or not to return to old position</param>
        ''' <param name="FigletFont">Figlet font to use in the text.</param>
        ''' <param name="color">A color that will be changed to.</param>
        ''' <param name="Vars">Variables to format the message before it's written.</param>
        Public Sub WriteFigletWhere(Text As String, Left As Integer, Top As Integer, [Return] As Boolean, FigletFont As FiggleFont, Color As ConsoleColor, ParamArray Vars() As Object)
            Try
                'Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                SetConsoleColor(New Color(Color))

                'Actually write
                WriteFigletWherePlain(Text, Left, Top, [Return], FigletFont, Vars)
            Catch ex As Exception When Not ex.GetType.Name = "ThreadAbortException"
                WStkTrc(ex)
                KernelError(KernelErrorLevel.C, False, 0, DoTranslation("There is a serious error when printing text."), ex)
            End Try
        End Sub

        ''' <summary>
        ''' Writes the figlet text with position support
        ''' </summary>
        ''' <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        ''' <param name="Left">Column number in console</param>
        ''' <param name="Top">Row number in console</param>
        ''' <param name="Return">Whether or not to return to old position</param>
        ''' <param name="FigletFont">Figlet font to use in the text.</param>
        ''' <param name="ForegroundColor">A foreground color that will be changed to.</param>
        ''' <param name="BackgroundColor">A background color that will be changed to.</param>
        ''' <param name="Vars">Variables to format the message before it's written.</param>
        Public Sub WriteFigletWhere(Text As String, Left As Integer, Top As Integer, [Return] As Boolean, FigletFont As FiggleFont, ForegroundColor As ConsoleColor, BackgroundColor As ConsoleColor, ParamArray Vars() As Object)
            Try
                'Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                SetConsoleColor(New Color(ForegroundColor))
                SetConsoleColor(New Color(BackgroundColor), True)

                'Actually write
                WriteFigletWherePlain(Text, Left, Top, [Return], FigletFont, Vars)
            Catch ex As Exception When Not ex.GetType.Name = "ThreadAbortException"
                WStkTrc(ex)
                KernelError(KernelErrorLevel.C, False, 0, DoTranslation("There is a serious error when printing text."), ex)
            End Try
        End Sub

        ''' <summary>
        ''' Writes the figlet text with position support
        ''' </summary>
        ''' <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        ''' <param name="Left">Column number in console</param>
        ''' <param name="Top">Row number in console</param>
        ''' <param name="Return">Whether or not to return to old position</param>
        ''' <param name="FigletFont">Figlet font to use in the text.</param>
        ''' <param name="Color">A color that will be changed to.</param>
        ''' <param name="Vars">Variables to format the message before it's written.</param>
        Public Sub WriteFigletWhere(Text As String, Left As Integer, Top As Integer, [Return] As Boolean, FigletFont As FiggleFont, Color As Color, ParamArray Vars() As Object)
            Try
                'Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                SetConsoleColor(Color)

                'Actually write
                WriteFigletWherePlain(Text, Left, Top, [Return], FigletFont, Vars)
            Catch ex As Exception When Not ex.GetType.Name = "ThreadAbortException"
                WStkTrc(ex)
                KernelError(KernelErrorLevel.C, False, 0, DoTranslation("There is a serious error when printing text."), ex)
            End Try
        End Sub

        ''' <summary>
        ''' Writes the figlet text with position support
        ''' </summary>
        ''' <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        ''' <param name="Left">Column number in console</param>
        ''' <param name="Top">Row number in console</param>
        ''' <param name="Return">Whether or not to return to old position</param>
        ''' <param name="FigletFont">Figlet font to use in the text.</param>
        ''' <param name="ForegroundColor">A foreground color that will be changed to.</param>
        ''' <param name="BackgroundColor">A background color that will be changed to.</param>
        ''' <param name="Vars">Variables to format the message before it's written.</param>
        Public Sub WriteFigletWhere(Text As String, Left As Integer, Top As Integer, [Return] As Boolean, FigletFont As FiggleFont, ForegroundColor As Color, BackgroundColor As Color, ParamArray Vars() As Object)
            Try
                'Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                SetConsoleColor(ForegroundColor)
                SetConsoleColor(BackgroundColor, True)

                'Actually write
                WriteFigletWherePlain(Text, Left, Top, [Return], FigletFont, Vars)
            Catch ex As Exception When Not ex.GetType.Name = "ThreadAbortException"
                WStkTrc(ex)
                KernelError(KernelErrorLevel.C, False, 0, DoTranslation("There is a serious error when printing text."), ex)
            End Try
        End Sub

    End Module
End Namespace

File: Kernel Simulator/Misc/Writers/FancyWriters/FigletWhereColor.cs
﻿
// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using ColorSeq;
using Figgle;
using KS.ConsoleBase.Colors;
using KS.Kernel;
using KS.Kernel.Debugging;
using KS.Languages;
using KS.Misc.Reflection;
using KS.Misc.Writers.ConsoleWriters;
using KS.Misc.Writers.WriterBase;

namespace KS.Misc.Writers.FancyWriters
{
    public static class FigletWhereColor
    {

        /// <summary>
        /// Writes the figlet text with position support
        /// </summary>
        /// <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        /// <param name="Left">Column number in console</param>
        /// <param name="Top">Row number in console</param>
        /// <param name="Return">Whether or not to return to old position</param>
        /// <param name="FigletFont">Figlet font to use in the text.</param>
        /// <param name="Vars">Variables to format the message before it's written.</param>
        public static void WriteFigletWherePlain(string Text, int Left, int Top, bool Return, FiggleFont FigletFont, params object[] Vars)
        {
            try
            {
                // Format string as needed
                if (!(Vars.Length == 0))
                    Text = StringManipulate.FormatString(Text, Vars);

                // Write the font
                Text = FigletFont.Render(Text);
                WriterPlainManager.currentPlain.WriteWherePlain(Text, Left, Top, Return, Vars);
            }
            catch (Exception ex) when (!(ex.GetType().Name == "ThreadInterruptedException"))
            {
                DebugWriter.WStkTrc(ex);
                KernelTools.KernelError(KernelErrorLevel.C, false, 0L, Translate.DoTranslation("There is a serious error when printing text."), ex);
            }
        }

        /// <summary>
        /// Writes the figlet text with position support
        /// </summary>
        /// <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        /// <param name="Left">Column number in console</param>
        /// <param name="Top">Row number in console</param>
        /// <param name="Return">Whether or not to return to old position</param>
        /// <param name="FigletFont">Figlet font to use in the text.</param>
        /// <param name="ColTypes">A type of colors that will be changed.</param>
        /// <param name="Vars">Variables to format the message before it's written.</param>
        public static void WriteFigletWhere(string Text, int Left, int Top, bool Return, FiggleFont FigletFont, ColorTools.ColTypes ColTypes, params object[] Vars)
        {
            try
            {
                // Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                ColorTools.SetConsoleColor(ColTypes);

                // Actually write
                WriteFigletWherePlain(Text, Left, Top, Return, FigletFont, Vars);
            }
            catch (Exception ex) when (!(ex.GetType().Name == "ThreadInterruptedException"))
            {
                DebugWriter.WStkTrc(ex);
                KernelTools.KernelError(KernelErrorLevel.C, false, 0L, Translate.DoTranslation("There is a serious error when printing text."), ex);
            }
        }

        /// <summary>
        /// Writes the figlet text with position support
        /// </summary>
        /// <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        /// <param name="Left">Column number in console</param>
        /// <param name="Top">Row number in console</param>
        /// <param name="Return">Whether or not to return to old position</param>
        /// <param name="FigletFont">Figlet font to use in the text.</param>
        /// <param name="colorTypeForeground">A type of colors that will be changed for the foreground color.</param>
        /// <param name="colorTypeBackground">A type of colors that will be changed for the background color.</param>
        /// <param name="Vars">Variables to format the message before it's written.</param>
        public static void WriteFigletWhere(string Text, int Left, int Top, bool Return, FiggleFont FigletFont, ColorTools.ColTypes colorTypeForeground, ColorTools.ColTypes colorTypeBackground, params object[] Vars)
        {
            try
            {
                // Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                ColorTools.SetConsoleColor(colorTypeForeground);
                ColorTools.SetConsoleColor(colorTypeBackground, true);

                // Actually write
                WriteFigletWherePlain(Text, Left, Top, Return, FigletFont, Vars);
            }
            catch (Exception ex) when (!(ex.GetType().Name == "ThreadInterruptedException"))
            {
                DebugWriter.WStkTrc(ex);
                KernelTools.KernelError(KernelErrorLevel.C, false, 0L, Translate.DoTranslation("There is a serious error when printing text."), ex);
            }
        }

        /// <summary>
        /// Writes the figlet text with position support
        /// </summary>
        /// <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        /// <param name="Left">Column number in console</param>
        /// <param name="Top">Row number in console</param>
        /// <param name="Return">Whether or not to return to old position</param>
        /// <param name="FigletFont">Figlet font to use in the text.</param>
        /// <param name="Color">A color that will be changed to.</param>
        /// <param name="Vars">Variables to format the message before it's written.</param>
        public static void WriteFigletWhere(string Text, int Left, int Top, bool Return, FiggleFont FigletFont, ConsoleColor Color, params object[] Vars)
        {
            try
            {
                // Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                ColorTools.SetConsoleColor(new Color((int)Color));

                // Actually write
                WriteFigletWherePlain(Text, Left, Top, Return, FigletFont, Vars);
            }
            catch (Exception ex) when (!(ex.GetType().Name == "ThreadInterruptedException"))
            {
                DebugWriter.WStkTrc(ex);
                KernelTools.KernelError(KernelErrorLevel.C, false, 0L, Translate.DoTranslation("There is a serious error when printing text."), ex);
            }
        }

        /// <summary>
        /// Writes the figlet text with position support
        /// </summary>
        /// <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        /// <param name="Left">Column number in console</param>
        /// <param name="Top">Row number in console</param>
        /// <param name="Return">Whether or not to return to old position</param>
        /// <param name="FigletFont">Figlet font to use in the text.</param>
        /// <param name="ForegroundColor">A foreground color that will be changed to.</param>
        /// <param name="BackgroundColor">A background color that will be changed to.</param>
        /// <param name="Vars">Variables to format the message before it's written.</param>
        public static void WriteFigletWhere(string Text, int Left, int Top, bool Return, FiggleFont FigletFont, ConsoleColor ForegroundColor, ConsoleColor BackgroundColor, params object[] Vars)
        {
            try
            {
                // Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                ColorTools.SetConsoleColor(new Color((int)ForegroundColor));
                ColorTools.SetConsoleColor(new Color((int)BackgroundColor), true);

                // Actually write
                WriteFigletWherePlain(Text, Left, Top, Return, FigletFont, Vars);
            }
            catch (Exception ex) when (!(ex.GetType().Name == "ThreadInterruptedException"))
            {
                DebugWriter.WStkTrc(ex);
                KernelTools.KernelError(KernelErrorLevel.C, false, 0L, Translate.DoTranslation("There is a serious error when printing text."), ex);
            }
        }

        /// <summary>
        /// Writes the figlet text with position support
        /// </summary>
        /// <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        /// <param name="Left">Column number in console</param>
        /// <param name="Top">Row number in console</param>
        /// <param name="Return">Whether or not to return to old position</param>
        /// <param name="FigletFont">Figlet font to use in the text.</param>
        /// <param name="Color">A color that will be changed to.</param>
        /// <param name="Vars">Variables to format the message before it's written.</param>
        public static void WriteFigletWhere(string Text, int Left, int Top, bool Return, FiggleFont FigletFont, Color Color, params object[] Vars)
        {
            try
            {
                // Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                ColorTools.SetConsoleColor(Color);

                // Actually write
                WriteFigletWherePlain(Text, Left, Top, Return, FigletFont, Vars);
            }
            catch (Exception ex) when (!(ex.GetType().Name == "ThreadInterruptedException"))
            {
                DebugWriter.WStkTrc(ex);
                KernelTools.KernelError(KernelErrorLevel.C, false, 0L, Translate.DoTranslation("There is a serious error when printing text."), ex);
            }
        }

        /// <summary>
        /// Writes the figlet text with position support
        /// </summary>
        /// <param name="Text">Text to be written. If nothing, the entire line is filled with the separator.</param>
        /// <param name="Left">Column number in console</param>
        /// <param name="Top">Row number in console</param>
        /// <param name="Return">Whether or not to return to old position</param>
        /// <param name="FigletFont">Figlet font to use in the text.</param>
        /// <param name="ForegroundColor">A foreground color that will be changed to.</param>
        /// <param name="BackgroundColor">A background color that will be changed to.</param>
        /// <param name="Vars">Variables to format the message before it's written.</param>
        public static void WriteFigletWhere(string Text, int Left, int Top, bool Return, FiggleFont FigletFont, Color ForegroundColor, Color BackgroundColor, params object[] Vars)
        {
            try
            {
                // Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
                ColorTools.SetConsoleColor(ForegroundColor);
                ColorTools.SetConsoleColor(BackgroundColor, true);

                // Actually write
                WriteFigletWherePlain(Text, Left, Top, Return, FigletFont, Vars);
            }
            catch (Exception ex) when (!(ex.GetType().Name == "ThreadInterruptedException"))
            {
                DebugWriter.WStkTrc(ex);
                KernelTools.KernelError(KernelErrorLevel.C, false, 0L, Translate.DoTranslation("There is a serious error when printing text."), ex);
            }
        }

    }
}
